aya/src/obj/btf/btf.rs

```diff
@@ -422,143 +404,129 @@ impl Btf {
-        for t in self.types().skip(1) {
-            let b = t.to_bytes();
-            buf.put(b.as_slice())
-        }
+        buf.extend(self.types.to_bytes());
         buf.put(self.strings.as_slice());
         buf
     }
 
-    pub(crate) fn fixup(
+    pub(crate) fn fixup_and_sanitize(
         &mut self,
         section_sizes: &HashMap<String, u64>,
         symbol_offsets: &HashMap<String, u64>,
+        features: &Features,
     ) -> Result<(), BtfError> {
-        let mut types = self.types.split_off(0);
-        for t in &mut types {
-            let kind = t.kind()?.unwrap_or_default();
+        let mut types = mem::take(&mut self.types);
+        for i in 0..types.types.len() {
+            let kind = types.types.get(i).unwrap().kind()?.unwrap_or_default();
             // datasec sizes aren't set by llvm
             // we need to fix them here before loading the btf to the kernel
-            match t {
-                BtfType::Ptr(mut ty) => {
-                    // Rust emits names for pointer types, which the kernel doesn't like
-                    // While I figure out if this needs fixing in the Kernel or LLVM, we'll
-                    // do a fixup here
-                    ty.name_off = 0;
+            match kind {
+                // Fixup PTR for Rust
+                // LLVM emits names for Rust pointer types, which the kernel doesn't like
+                // While I figure out if this needs fixing in the Kernel or LLVM, we'll
+                // do a fixup here
+                BtfKind::Ptr => {
+                    if let Some(BtfType::Ptr(mut ty)) = types.types.get_mut(i) {
+                        ty.name_off = 0;
+                    }
                 }
-                BtfType::DataSec(mut ty, data) => {
-                    // Start DataSec Fixups
-                    let sec_name = self.string_at(ty.name_off)?;
-                    let name = sec_name.to_string();
-                    // There are cases when the compiler does indeed populate the
-                    // size. If we hit this case, push to the types vector and
-                    // continue
-                    if unsafe { ty.__bindgen_anon_1.size > 0 } {
-                        debug!("{} {}: fixup not required", kind, name);
-                        continue;
+                // Sanitize VAR if they are not supported
+                BtfKind::Var if !features.btf_datasec => {
+                    if let Some(BtfType::Var(ty, _)) = types.types.get(i) {
+                        types.types[i] = BtfType::new_int(ty.name_off, 1, 0, 0);
                     }
+                }
+                // Sanitize DATASEC if they are not supported
+                BtfKind::DataSec if !features.btf_datasec => {
+                    if let Some(BtfType::DataSec(ty, data)) = types.types.get(i) {
+                        debug!("{}: not supported. replacing with STRUCT", kind);
+                        let mut members = vec![];
```

Comment:
iiuc you're just changing the offset here, so you could clone data and then just mutate the offset in place like we do for the BtfType::DataSec arm

Reply:
we're actually changing between types here. we could use mem::transmute since the layout is the same, but this is safer